host/switchless_manager.c

```diff
@@ -0,0 +1,126 @@
+/* Copyright (c) Microsoft Corporation.  All rights reserved.
+ * Licensed under the MIT License. */
+
+#include "switchless_manager.h"
+#include <openenclave/internal/calls.h>
+#include "hostthread.h"
+#include "sgx/enclave.h"
+
+static oe_thread_return_t _enclave_worker_thread(oe_thread_arg_t arg)
+{
+    oe_enclave_t* enclave = (oe_enclave_t*)arg;
+
+    oe_ecall(
+        enclave,
+        OE_ECALL_LAUNCH_ENCLAVE_WORKER,
+        (uint64_t)(&enclave->switchless_manager.switchless),
+        NULL);
+
+    return OE_THREAD_RETURN_VAL;
+} /* _enclave_worker_thread */
+
+void oe_switchless_manager_init(oe_enclave_t* enclave)
+{
+    enclave->switchless_manager.switchless.state = OE_SWITCHLESS_STATE_STOPPED;
+    oe_lockless_queue_init(
+        &(enclave->switchless_manager.switchless.ecall_queue));
+} /* oe_switchless_manager_init */
+
+oe_result_t oe_switchless_manager_startup(oe_enclave_t* enclave)
+{
+    oe_result_t result = OE_OK;
+    oe_switchless_state_t state = OE_SWITCHLESS_STATE_STOPPED;
+
+    /* check the current state:
+       - if it is stopped then change to starting */
+#ifdef _MSC_VER
+    state = (oe_switchless_state_t)_InterlockedCompareExchange(
+        &(enclave->switchless_manager.switchless.state),
+        OE_SWITCHLESS_STATE_STARTING,
+        OE_SWITCHLESS_STATE_STOPPED);
+#elif defined __GNUC__
+    __atomic_compare_exchange_n(
+        &(enclave->switchless_manager.switchless.state),
+        &state,
+        OE_SWITCHLESS_STATE_STARTING,
+        1,
+        __ATOMIC_ACQ_REL,
+        __ATOMIC_ACQUIRE);
+#endif /* _MSC_VER or __GNUC__ */
+
+    if (OE_SWITCHLESS_STATE_STOPPED == state)
+    {
+        if (0 == oe_thread_create(
+                     &(enclave->switchless_manager.enclave_worker),
+                     _enclave_worker_thread,
+                     enclave))
+        {
+            /* change the state to running */
+#ifdef _MSC_VER
+            _InterlockedExchange(
+                &(enclave->switchless_manager.switchless.state),
+                OE_SWITCHLESS_STATE_RUNNING);
+#elif defined __GNUC__
+            __atomic_exchange_n(
+                &(enclave->switchless_manager.switchless.state),
+                OE_SWITCHLESS_STATE_RUNNING,
+                __ATOMIC_RELEASE);
+#endif /* _MSC_VER or __GNUC__ */
+        }
+        else
+        {
+            result = OE_FAILURE;
+        }
+    }
+    else
+    {
+        /* invalid state */
+        result = OE_FAILURE;
```

Comment:
If two threads simultaneously make a switchless ecall, then they might end up simultaneously calling  startup. As written currently, one of the threads will create the worker thread, whereas the other would return OE_FAILURE. This seems wrong.

Reply:
I've added code to ensure that only one thread can perform the startup.  If more than one thread calls startup at the same time, the second thread will wait until the first thread completes the startup and then return success.